ktor-client/ktor-client-core/common/src/io/ktor/client/features/NeedRetry.kt

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2014-2019 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.client.features
+
+import io.ktor.client.HttpClient
+import io.ktor.client.HttpClientConfig
+import io.ktor.client.request.HttpRequestBuilder
+import io.ktor.client.request.takeFrom
+import io.ktor.client.response.HttpReceivePipeline
+import io.ktor.client.response.HttpResponse
+import io.ktor.util.AttributeKey
+
+/**
+ * Need retry condition checker
+ *
+ * You could return true to make [HttpClient] retry the request
+ */
+typealias RetryCondition = suspend (requestBuilder: HttpRequestBuilder, response: HttpResponse) -> Boolean
+
+/**
+ * Need retry feature for [HttpClient]
+ *
+ * @property retryHandlers: list of retry conditions to check before deciding to retry a request
+ */
+class NeedRetry(
+    private val retryHandlers: List<RetryCondition>
+) {
+
+    /**
+     * [NeedRetry] configuration.
+     */
+    class Config {
+        internal val retryHandlers: MutableList<RetryCondition> = mutableListOf()
+
+        /**
+         * Add [RetryCondition].
+         * Last added handler executes first.
+         */
+        fun retryCondition(block: RetryCondition) {
+            retryHandlers += block
+        }
+    }
+
+    companion object : HttpClientFeature<Config, NeedRetry> {
+        override val key: AttributeKey<NeedRetry> = AttributeKey("NeedRetry")
+
+        override fun prepare(block: Config.() -> Unit): NeedRetry {
+            val config = Config().apply(block)
+
+            config.retryHandlers.reversed()
```

Comment:
Just like in [HttpCallValidator](https://github.com/ktorio/ktor/blob/master/ktor-client/ktor-client-core/common/src/io/ktor/client/features/HttpCallValidator.kt#L72-L73) this is a useless call. The function `reversed` does NOT modify the `MutableList` but returns a new `List`. So either the return value needs to be used or replace it with `reverse`. See the following for more details: https://github.com/ktorio/ktor/issues/1345

Reply:
agreed :) thank you

Reply:
This isn't solved yet. You just removed the reversed call. What about your comment in line 39 which says:

> Last added handler executes first

Can you confirm this is the case? Maybe with a test?

Reply:
This still needs to be fixed.